### PR TITLE
[distill]add distillation losses

### PR DESCRIPTION
**Pull request by littletomatodonkey**
1. add combined loss, which is a combination of many loss function
2. add distillation loss, which can be used to parser model output
3. add basic loss, which is used to calculate loss for tensros
4. add unit-test for all losses

### CODE REVIEW

File: paddleslim/dygraph/dist/losses/basic_loss.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddle.nn import L1Loss
+from paddle.nn import MSELoss as L2Loss
+from paddle.nn import SmoothL1Loss
+
+
+class CELoss(nn.Layer):
+    def __init__(self, epsilon=None, label_act="softmax"):
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/losses/basic_loss.py]: add comments

File: paddleslim/dygraph/dist/losses/basic_loss.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+            one_hot_target = target
+        soft_target = F.label_smooth(one_hot_target, epsilon=self.epsilon)
+        soft_target = paddle.reshape(soft_target, shape=[-1, class_num])
+        return soft_target
+
+    def forward(self, x, label):
+        if self.epsilon is not None:
+            class_num = x.shape[-1]
+            label = self._labelsmoothing(label, class_num)
+            x = -F.log_softmax(x, axis=-1)
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/losses/basic_loss.py]: 这里对axis做了很强的限定，对非last dim的会出错。但没有地方告诉用户

File: paddleslim/dygraph/dist/losses/basic_loss.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        loss = loss.mean()
+        return loss
+
+
+class DMLLoss(nn.Layer):
+    """
+    DMLLoss
+    """
+
+    def __init__(self, act=None):
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/losses/basic_loss.py]: 同上，add comments

File: paddleslim/dygraph/dist/losses/distillation_loss.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    def forward(self, predicts, batch):
+        loss_dict = dict()
+        for idx, pair in enumerate(self.model_name_pairs):
+            out1 = predicts[pair[0]]
+            out2 = predicts[pair[1]]
+            if self.key is not None:
+                out1 = out1[self.key]
+                out2 = out2[self.key]
+            loss_dict["{}_{}_{}_{}".format(self.name, pair[0], pair[1],
```
Comment (qingqing01) [inline paddleslim/dygraph/dist/losses/distillation_loss.py:58]: 这里的key和 https://github.com/PaddlePaddle/PaddleSlim/pull/799/files  这个PR有关联吗？

File: paddleslim/dygraph/dist/losses/__init__.py [hunk 5]
```diff
@@ -11,3 +11,59 @@
...
+
+from .distillation_loss import DistillationDistanceLoss
+from .distillation_loss import DistillationDMLLoss
+from .distillation_loss import DistillationRKDLoss
+
+
+class CombinedLoss(nn.Layer):
+    """
+    CombinedLoss:
+        a combionation of loss function
```
Comment (wanghaoshuang) [inline paddleslim/dygraph/dist/losses/__init__.py]: `combionation` -> `combination`

File: paddleslim/dygraph/dist/losses/__init__.py [hunk 6]
```diff
@@ -11,3 +11,59 @@
...
+            'operator config should be a list')
+        for config in loss_config_list:
+            assert isinstance(config,
+                              dict) and len(config) == 1, "yaml format error"
+            name = list(config)[0]
+            param = config[name]
+            assert "weight" in param, "weight must be in param, but param just contains {}".format(
+                param.keys())
+            self.loss_weight.append(param.pop("weight"))
+            self.loss_func.append(eval(name)(**param))
```
Comment (wanghaoshuang) [inline paddleslim/dygraph/dist/losses/__init__.py:67]: 如果 `name` 不是预定义的loss，会报`* is not defined`的错误，要提示用户可选的loss有哪些？
另外，用户还可能将`name`指定为`basic_loss`

File: paddleslim/dygraph/dist/losses/distillation_loss.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    def __init__(self, model_name_pairs=[], key=None, name="loss_rkd"):
+        super().__init__()
+        self.model_name_pairs = model_name_pairs
+        self.key = key
+
+        self.rkd_angle_loss_func = RKdAngle()
+        self.rkd_dist_func = RkdDistance()
+        self.name = name
+
+    def forward(self, predicts, batch):
```
Comment (wanghaoshuang) [inline paddleslim/dygraph/dist/losses/distillation_loss.py:123]: 这里的batch是什么意思，我看没有地方会用到这个输入？

File: paddleslim/dygraph/dist/losses/distillation_loss.py (hunk 4 above)
Comment (littletomatodonkey) [reply paddleslim/dygraph/dist/losses/distillation_loss.py:58]: 799中输出的内容应该为dict，计算loss的时候会根据key去取不同submodel的feature map，应该是有关系的~

File: paddleslim/dygraph/dist/losses/distillation_loss.py (hunk 7 above)
Comment (littletomatodonkey) [reply paddleslim/dygraph/dist/losses/distillation_loss.py:123]: 因为算蒸馏loss的时候，可能会根据gt去判断最终用于计算loss的样本，比如说难样本挖掘等，这里为了统一接口，在训练时就把所有的输入都传入进来了(网络的输出，即predicts，以及dataloader的输出，即batch)，按需解析~

File: paddleslim/dygraph/dist/losses/__init__.py (hunk 6 above)
Comment (littletomatodonkey) [reply paddleslim/dygraph/dist/losses/__init__.py:67]: 好的，我添加一下提示，另外，basic loss也是可以的，只要输入符合basic loss的接口定义即可（一般为x、y 2个tensor）

File: paddleslim/dygraph/dist/losses/basic_loss.py (hunk 1 above)
Comment (littletomatodonkey) [reply paddleslim/dygraph/dist/losses/basic_loss.py]: done, thx

File: paddleslim/dygraph/dist/losses/basic_loss.py (hunk 2 above)
Comment (littletomatodonkey) [reply paddleslim/dygraph/dist/losses/basic_loss.py]: 把axis参数抽出来了，多谢提醒~

File: paddleslim/dygraph/dist/losses/basic_loss.py (hunk 3 above)
Comment (littletomatodonkey) [reply paddleslim/dygraph/dist/losses/basic_loss.py]: done， thx

File: paddleslim/dygraph/dist/losses/__init__.py (hunk 5 above)
Comment (littletomatodonkey) [reply paddleslim/dygraph/dist/losses/__init__.py]: done, thx

**Review by wanghaoshuang — APPROVED** [review]
